common/thrift_client_pool.h

```diff
@@ -307,24 +307,30 @@ class ThriftClientPool {
     }
 
     void cleanupStaleChannels(const folly::SocketAddress& addr) {
-      // cleanup stale entries if it hasn't been done for a period of time.
       auto now = time(nullptr);
-      if (last_cleanup_time_ + FLAGS_channel_cleanup_min_interval_seconds
-          < now) {
-        last_cleanup_time_ = now;
-        auto itor = channels_.find(addr);
-        int n = 0;
-        while (itor != channels_.end() &&
-               n++ < FLAGS_channel_max_checking_size) {
-          // We don't cleanup !good() live channels here. Otherwise we
-          // will need to upgrade it to a shared_ptr. We expect clients
-          // won't keep a !good() channels for a long period of time.
-          if (itor->second.first.use_count() == 0) {
-            itor = channels_.erase(itor);
-          } else {
-            ++itor;
-          }
-        }
+      // skip cleanup if it was done recently
+      if (now < last_cleanup_time_ + FLAGS_channel_cleanup_min_interval_seconds) {
+	return;
+      }
+
+      last_cleanup_time_ = now;
+      auto itor = channels_.find(addr);
+      int n = 0;
+      while (itor != channels_.end() &&
+	     n++ < FLAGS_channel_max_checking_size) {
+	auto c = itor->second.first.lock();
```

Comment:
Is the lock here to ensure that the weak ptr is not released while it's status is checked in the next few lines?

Reply:
Oh lock() converts the weak pointer to a shared pointer: https://en.cppreference.com/w/cpp/memory/weak_ptr/lock
And we need to convert to a shared pointer to be able to access the underlying object.

Reply:
this creates a shared_ptr every time, for `FLAGS_channel_max_checking_size` number of times, is that indented?

Reply:
Yes, it's intended. itor will be advanced either via erase() or ++ below. So it creates a shared_ptr for different channel each time.

Reply:
oh I misread -- I thought this function only cleans the channel for the specified `addr`. It seems it will clean-up up to `FLAGS_channel_max_checking_size` channels? in the common case I think that may be wasted work assuming connections are healthy?

Not introduce by this PR though, so just to understand the original motivation.

Reply:
If a connection is health, the client (thrift_router, etc) of thrift_client_pool will cache the returned shared_ptr so that it can be reused.  In this case, the weak_ptr will be able to lock() into a shared_ptr, so this function won't do anything to it.

Reply:
also, created https://github.com/pinterest/rocksplicator/pull/565 as an alternative to this approach, i.e. since we already know when a channel is bad, we can close it right before creating a new one for a destination address, wdyt?

Reply:
Feel free to try it out. I can abandon this one if #565 works.
One subtle difference is that #565 only close connections that are being requested. This PR will clean up bad connections to any destinations which may not be requested.